backend/de.metas.business/src/main/java/de/metas/invoice/service/MInvoicePOCopyRecordSupport.java

```diff
@@ -1,29 +1,80 @@
 /**
- * 
+ *
  */
 package de.metas.invoice.service;
 
-import java.util.List;
-
+import com.google.common.collect.ImmutableList;
+import de.metas.invoice.InvoiceId;
+import de.metas.invoice.InvoiceLineId;
+import de.metas.invoice.detail.InvoiceDetailCloneMapper;
+import de.metas.invoice.detail.InvoiceWithDetailsService;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
 import org.adempiere.model.CopyRecordSupportTableInfo;
 import org.adempiere.model.GeneralCopyRecordSupport;
+import org.adempiere.model.InterfaceWrapperHelper;
+import org.compiere.SpringContextHolder;
+import org.compiere.model.I_C_Invoice;
 import org.compiere.model.I_C_InvoiceLine;
 import org.compiere.model.PO;
 
-import com.google.common.collect.ImmutableList;
+import javax.annotation.Nullable;
+import java.util.List;
 
-/**
- * @author Cristina Ghita, METAS.RO
- *         version for copy an invoice
- */
 public class MInvoicePOCopyRecordSupport extends GeneralCopyRecordSupport
 {
+	final InvoiceWithDetailsService invoiceWithDetailsService = SpringContextHolder.instance.getBean(InvoiceWithDetailsService.class);
+
 	@Override
 	public List<CopyRecordSupportTableInfo> getSuggestedChildren(final PO po, final List<CopyRecordSupportTableInfo> suggestedChildren)
 	{
 		return super.getSuggestedChildren(po, suggestedChildren)
 				.stream()
-				.filter(childTableInfo -> I_C_InvoiceLine.Table_Name.equals(childTableInfo.getTableName()))
+				.filter(MInvoicePOCopyRecordSupport::isSuggestedChildren)
 				.collect(ImmutableList.toImmutableList());
 	}
+
+	@Override
+	protected void onRecordAndChildrenCopied(final PO to, final PO from)
+	{
+		final I_C_Invoice fromInvoice = InterfaceWrapperHelper.create(from, I_C_Invoice.class);
+		final I_C_Invoice toInvoice = InterfaceWrapperHelper.create(to, I_C_Invoice.class);
+		onRecordAndChildrenCopied(fromInvoice, toInvoice);
+	}
+
+	private void onRecordAndChildrenCopied(final I_C_Invoice fromInvoice, final I_C_Invoice targetInvoice)
+	{
+		invoiceWithDetailsService.copyDetailsToClone(InvoiceDetailCloneMapperImpl.builder()
+															 .originalInvoiceId(InvoiceId.ofRepoId(fromInvoice.getC_Invoice_ID()))
+															 .targetInvoiceId(InvoiceId.ofRepoId(targetInvoice.getC_Invoice_ID()))
+															 .clonedInvoiceLinesInfo(MInvoiceLinePOCopyRecordSupport.getClonedInvoiceLinesInfo(targetInvoice))
+															 .build());
+	}
+
+	private static boolean isSuggestedChildren(@NonNull final CopyRecordSupportTableInfo childTableInfo)
+	{
+		return I_C_InvoiceLine.Table_Name.equals(childTableInfo.getTableName());
+	}
+
+	@Builder
+	private static class InvoiceDetailCloneMapperImpl implements InvoiceDetailCloneMapper
+	{
+		@NonNull @Getter private final InvoiceId originalInvoiceId;
+		@NonNull @Getter private final InvoiceId targetInvoiceId;
+		@Nullable private final MInvoiceLinePOCopyRecordSupport.ClonedInvoiceLinesInfo clonedInvoiceLinesInfo;
```

Comment:
why allowing this to be nullable and just later on getTargetInvoiceLineId failing if null?

Reply:
Because `org.adempiere.model.InterfaceWrapperHelper#getDynAttribute` returns nullable
In case of an invoice without lines, `de.metas.invoice.detail.InvoiceDetailCloneMapper#clonedInvoiceLinesInfo` is going to be null, but in that case we won't call `de.metas.invoice.detail.InvoiceDetailCloneMapper#getTargetInvoiceLineId`